### PR TITLE
Use SentimentServlet to retrieve video data and analyze sentiment.

### PR DESCRIPTION
**Pull request by kayefung**
When a user inputs a link to a YT video and click the analyze button the SentimentServlet.java will call CommentService.java and Caption.java to get the comments and captions of the inputted video. The text will get analyzed in calculateSentimentScore(). The score can be retrieved with a GET request (coming in an future PR)

I also mentioned in the pod sync that I'm having problems making multiple calls to the NLP API, for now we're going to put the captions and comments together into one String and analyze its sentiment as a whole.

### CODE REVIEW

File: src/main/java/com/google/sps/servlets/SentimentServlet.java [hunk 1]
```diff
@@ -19,71 +22,79 @@
...
+    String comments = String.join(". ", commentsList);
+
+    String captions = captionService.getCaptionFromId(videoId);
 
-    if (comments.isEmpty()) {
-      response.sendError(HttpServletResponse.SC_BAD_REQUEST, EMPTY_STRING_ERROR);
+    if (comments.isEmpty() && captions.isEmpty()) {
+      response.sendError(HttpServletResponse.SC_INTERNAL_SERVER_ERROR, NO_DATA_ERROR);
       return;
     }
```
Comment (juanalonsocruz) [inline src/main/java/com/google/sps/servlets/SentimentServlet.java:77]: Instead of any error for no data, why not just tell the client that as part of the response (VideoAnalysis)? Maybe add another field to VideoAnalysis that conveys this information.

File: src/main/webapp/scripts/analyze.js [hunk 2]
```diff
@@ -0,0 +1,7 @@
+function analyzeVideo() {
+  const videoId = document.getElementById("video-url").value.split('v=')[1];
+  fetch('/sentiment', {
+    method: 'POST',
+    body: videoId
+  });
+}
```
Comment (juanalonsocruz) [inline src/main/webapp/scripts/analyze.js:7]: Keep JS changes to a different PR unless it is required due to breaking/changing a previous feature..

File: src/test/java/com/google/sps/SentimentServletTest.java [hunk 3]
```diff
@@ -53,97 +61,159 @@ public void setUp() throws IOException {
...
         .thenReturn(sentimentMock);
     when(sentimentMock.getScore()).thenReturn(SCORE);
     doNothing().when(languageServiceMock).close();
 
+    sentimentServlet.doPost(requestMock, responseSpy);
+
     sentimentServlet.doGet(requestMock, responseSpy);
 
-    Assert.assertEquals(stringWriter.toString(), "{\"commentScore\":0.05}\n");
+    Assert.assertEquals(stringWriter.toString(), "{\"score\":0.05}\n");
```
Comment (juanalonsocruz) [inline src/test/java/com/google/sps/SentimentServletTest.java:82]: Could you build the expected string inside the unit test using the SCORE variable. Its not obvious that 0.05 is the value of SCORE by just reading the unit test.

File: src/main/java/com/google/sps/servlets/SentimentServlet.java [hunk 4]
```diff
@@ -19,71 +22,79 @@
...
     languageService = LanguageServiceClient.create();
+    captionService = new Caption();
+    commentService = new CommentService();
   }
 
   @Override
   public void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException {
-    String comments = "";
-    Float commentsScore = null;
+    String json = new Gson().toJson(videoAnalysis);
```
[the remark this answers is not in the document]
Comment (chenzizhao) [reply src/main/java/com/google/sps/servlets/SentimentServlet.java:44]: Just to add on to the suggestion: I experienced some difficulty to add a body to a GET request (in JavaScript) and I opted for query string instead..

https://stackoverflow.com/questions/978061/http-get-with-request-body#:~:text=Yes.,semantic%20meaning%20to%20the%20request.

File: src/test/java/com/google/sps/SentimentServletTest.java [hunk 5]
```diff
@@ -50,100 +57,156 @@ public void setUp() throws IOException {
...
+    doNothing().when(languageServiceMock).close();
 
     sentimentServlet.doGet(requestMock, responseSpy);
 
+    String expected = "{\"score\":" + df.format(SCORE) + ",\"dataAvailable\":true}\n";
+
     verify(responseSpy).sendError(HttpServletResponse.SC_INTERNAL_SERVER_ERROR,
-        "Language service client failed.");
+        "Comments could not be retrieved.");
+    Assert.assertEquals(expected, stringWriter.toString());
```
Comment (chenzizhao) [inline src/test/java/com/google/sps/SentimentServletTest.java:190]: Is it okay to send both the error and the value to the front end? For example, if I want to get the score in the front end, do I find it in fetch..then.. or fetch..catch.. ?

Comment (chenzizhao) [reply src/test/java/com/google/sps/SentimentServletTest.java:190]: nvm now I know you are reading the status and the message

File: src/main/java/com/google/sps/servlets/SentimentServlet.java [hunk 6]
```diff
@@ -19,71 +22,74 @@
...
+  private final String COMMENTS_FAILED_ERROR = "Comments could not be retrieved.";
   private final String NLP_API_ERROR = "Language service client failed.";
+  private final String NO_DATA_ERROR = "No comments or captions available to analyze.";
+  private final String READER_ERROR = "Reading video ID failed.";
 
-  public SentimentServlet() throws IOException {
+  public SentimentServlet() throws IOException, GeneralSecurityException {
     languageService = LanguageServiceClient.create();
+    captionService = new Caption();
+    commentService = new CommentService();
```
[the remark this answers is not in the document]
Comment (kayefung) [reply src/main/java/com/google/sps/servlets/SentimentServlet.java:38]: It made injecting mocks during testing easier (if I put it in the constructor and avoid calling ```new SentimentServlet()``` then it won't replace the mocks with real instances of the classes). I definitely think it looks weird though lol

[the remark this answers is not in the document]
Comment (chenzizhao) [reply src/main/java/com/google/sps/servlets/SentimentServlet.java:38]: ooook that makes sense, thanks!
